### PR TITLE
Upgrade transforms without upgrading the pipelines

### PR DESCRIPTION
**Pull request by chamikaramj**
This uses the Beam Transform Service [1] to support upgrade Beam Java transforms used by pipelines without upgrading the Beam version used by the pipelines. 

A similar solution can be introduced for other SDKs in the future.

For more details please see https://github.com/apache/beam/issues/27943.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CombineTranslation.java [hunk 1]
```diff
@@ -62,6 +64,11 @@ private CombinePerKeyPayloadTranslator() {}
 
     @Override
     public String getUrn(Combine.PerKey<?, ?, ?> transform) {
+      return getUrn();
```
[the remark this answers is not in the document]
Comment (chamikaramj) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CombineTranslation.java]: Yeah, moved.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CombineTranslation.java [hunk 2]
```diff
@@ -84,6 +91,16 @@ public FunctionSpec translate(
       }
     }
 
+    @Override
+    public @Nullable Row toConfigRow(PTransform<?, ?> pTransform) {
```
[the remark this answers is not in the document]
Comment (chamikaramj) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CombineTranslation.java]: I don't think these need  a separate interface but good point regarding providing default implementations. Made "toConfigRow" and "fromConfigRow" default methods of the interface so that sub-classes can choose to not implement them.

File: sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/ExpansionService.java [hunk 3]
```diff
@@ -178,6 +184,65 @@ public List<String> getDependencies(
...
+          TransformProvider transformProvider =
+              spec -> {
+                try {
+                  ExternalConfigurationPayload payload =
+                      ExternalConfigurationPayload.parseFrom(spec.getPayload());
+                  Row configRow =
+                      RowCoder.of(SchemaTranslation.schemaFromProto(payload.getSchema()))
+                          .decode(new ByteArrayInputStream(payload.getPayload().toByteArray()));
+                  @Nullable PTransform transformFromRow = translator.fromConfigRow(configRow);
+                  if (transformFromRow != null) {
```
[the remark this answers is not in the document]
Comment (chamikaramj) [reply sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/ExpansionService.java:228]: Now these methods return not nullable objects and provide default implementations that throw "UnsupportedOperationException".

**Review by chamikaramj — COMMENTED** [review]
Thanks Robert. PTAL.

**Comment by chamikaramj** [comment]
Also confirmed that this work for Dataflow by running a pipeline that upgrades the Create transform to a version available in an expansion service.

https://console.cloud.google.com/dataflow/jobs/us-central1/2023-09-12_15_32_47-10327070264136367924?project=apache-beam-testing

cc: @kennknowles

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java [hunk 4]
```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
...
+    List<String> urnsToOverride = externalTranslationOptions.getTransformsToOverride();
+    if (urnsToOverride.size() > 0) {
+      // We use PTransformPayloadTranslators and the Transform Service to re-generate the pipeline
+      // proto components for the updated transform and update the pipeline proto.
+      Map<String, PTransform> transforms = res.getComponents().getTransformsMap();
+      List<String> alreadyCheckedURns = new ArrayList<>();
+      for (Entry<String, PTransform> entry : transforms.entrySet()) {
+        String urn = entry.getValue().getSpec().getUrn();
+        if (!alreadyCheckedURns.contains(urn) && urnsToOverride.contains(urn)) {
+          alreadyCheckedURns.add(urn);
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java:136]: Won't this only override one transform of each type in the graph? (Seems worth a test too.)

Comment (robertwb) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java:136]: Oh, I see, for each transform you're then checking for any other transforms with that URN. I think it'd be cleaner to loop over URNs to translate rather than over transforms.entrySet() to get the set of suitable URNs. .

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java [hunk 5]
```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
...
+
+    // Java expansion serivice able to identify and expand transforms that includes the construction
+    // config provided here.
+    ExternalTransforms.ExternalConfigurationPayload payload =
+        ExternalTransforms.ExternalConfigurationPayload.newBuilder()
+            .setSchema(SchemaTranslation.schemaToProto(configRow.getSchema(), true))
+            .setPayload(outputStream.toByteString())
+            .build();
+
+    String serviceAddress = null;
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java:227]: Start and connect to the service once and pass it in as an option rather than doing so for every transform individually.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java [hunk 6]
```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
...
+                      org.apache.beam.sdk.transforms.PTransform<? super PInput, POutput>>>
+              appliedPTransforms =
+                  findAppliedPTransforms(
+                      urn, pipeline, KnownTransformPayloadTranslator.KNOWN_PAYLOAD_TRANSLATORS);
+          for (AppliedPTransform<
+                  PInput,
+                  POutput,
+                  org.apache.beam.sdk.transforms.PTransform<? super PInput, POutput>>
+              appliedPTransform : appliedPTransforms) {
+            TransformPayloadTranslator<
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java:151]: Rather than re-look this up, I still think it'd be cleaner to attach the Row to the PTransform itself during translation (at the same time you're attaching the URN and doing the translation, so you already have a PayloadTranslator in hand), and then just walk over the graph looking for URNs that you need to swap out. 

You can stick the config row (and its schema) in https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto#L207

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java [hunk 7]
```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
...
+              appliedPTransform.getInputs(),
+              newOutputs,
+              externalTransform,
+              externalTransform.getResourceHints(),
+              appliedPTransform.getPipeline());
+      SdkComponents updatedComponents =
+          SdkComponents.create(
+              runnerAPIpipeline.getComponents(), runnerAPIpipeline.getRequirementsList());
+      String updatedTransformId =
+          updatedComponents.registerPTransform(updatedAppliedPTransform, Collections.emptyList());
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java:328]: `Collections.emptyList()` doesn't seem correct if this is a composite transform.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java [hunk 8]
```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
...
+      RunnerApi.Components updatedRunnerApiComponents = updatedComponents.toComponents();
+
+      // Recording input updates to the transforms to refer to the upgraded transform instead of the
+      // old one.
+      // Also recording the newly generated id of the old (overridden) transform in the
+      // updatedRunnerApiComponents.
+      Map<String, Map<String, String>> transformInputUpdates = new HashMap<>();
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java:335]: Could you document what `transformInputUpdates` contains? String isn't very descriptive.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java [hunk 9]
```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
...
+                }
+              });
+      // There should be only one recorded old (upgraded) transform.
+      if (oldTransformIds.size() != 1) {
+        throw new IOException(
+            "Expected exactly one transform to be updated by "
+                + oldTransformIds.size()
+                + " were updated.");
+      }
+      String oldTransformId = oldTransformIds.get(0);
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java:365]: Shouldn't we already know exactly what transform we're updating? This seems a round-about way of getting it, and will break if a transform doesn't have outputs.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java [hunk 10]
```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
...
+
+      // Updated list of root transforms (in case a root was upgraded).
+      List<String> updaterRootTransformIds = new ArrayList<>();
+      updaterRootTransformIds.addAll(runnerAPIpipeline.getRootTransformIdsList());
+      if (updaterRootTransformIds.contains(oldTransformId)) {
+        updaterRootTransformIds.remove(oldTransformId);
+        updaterRootTransformIds.add(updatedTransformId);
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java:372]: I suppose the order does not matter here?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java [hunk 11]
```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
...
+
+      // After expansion some transforms might still refer to the output of the already overridden
+      // transform as their input.
+      // Such inputs have to be overridden to use the output of the new upgraded transform.
+      Map<String, String> inputReplacements = new HashMap<>();
+
+      // Will contain the outputs of the upgraded transform.
+      Map<TupleTag<?>, PCollection<?>> newOutputs = new HashMap<>();
+
+      if (output instanceof PCollectionTuple) {
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java:281]: This code deals with both the Pipeline object and the Pipeline proto, and it's a bit hard to follow where the source of truth is. My personal preference would be to simply act on the protos consuming the expansion service response directly. Honestly, I don't think that'd be any more complicated than what you're already having to do here already.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java [hunk 12]
```diff
@@ -508,14 +510,56 @@ static RunnerApi.PTransform.Builder translateAppliedPTransform(
...
+     */
+    default String getUrn(T transform) {
+      return getUrn();
+    }
+
+    /**
+     * Translates the given transform represented by the provided {@code AppliedPTransform} to a
+     * {@code FunctionSpec} with a URN and a payload.
+     */
     @Nullable
```
[the remark this answers is not in the document]
Comment (chamikaramj) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java:589]: Done.

**Review by chamikaramj — COMMENTED** [review]
Thanks Robert. PTAL.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java [hunk 13]
```diff
@@ -461,6 +475,38 @@ public RunnerApi.PTransform translate(
...
+        LOG.warn(
+            "Could not attach the config row for transform "
+                + appliedPTransform.getTransform().getName()
+                + ": "
+                + e);
+        // Ignoring the error and continuing with the translation since attaching config rows is
+        // optional.
+      }
+      if (configRow != null) {
+        ByteStringOutputStream rowOutputStream = new ByteStringOutputStream();
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java:494]: I think there's a CoderUtils that does this for you.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java [hunk 14]
```diff
@@ -461,6 +475,38 @@ public RunnerApi.PTransform translate(
...
+        try {
+          RowCoder.of(configRow.getSchema()).encode(configRow, rowOutputStream);
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+        transformBuilder.putAnnotations(CONFIG_ROW_KEY, rowOutputStream.toByteString());
+
+        ByteStringOutputStream schemaOutputStream = new ByteStringOutputStream();
+        try (ObjectOutputStream schemaObjOut = new ObjectOutputStream(schemaOutputStream)) {
+          schemaObjOut.writeObject(SchemaTranslation.schemaToProto(configRow.getSchema(), true));
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java:504]: SchemaTranslation.schemaToProto is a Proto object. Just use `toByteArray` to serialize this proto. You can use ByteString.copyFrom to go from byte[] to ByteString and not have to deal with streams at all.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+  }
+
+  private <
+          InputT extends PInput,
+          OutputT extends POutput,
+          TransformT extends org.apache.beam.sdk.transforms.PTransform<InputT, OutputT>>
+      RunnerApi.Pipeline updateTransformViaTransformService(
+          RunnerApi.Pipeline runnerAPIpipeline,
+          String transformId,
+          Endpoints.ApiServiceDescriptor transformServiceEndpoint)
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:155]: Perhaps create the client once and pass it here, rather than passing the endpoint and re-connecting it every time?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+   *
+   * @param pipeline the pipeline proto.
+   * @param urnsToOverride URNs of the transforms to be overridden.
+   * @param options options for determining the transform service to use.
+   * @return pipelines with transforms upgraded using the Transform Service.
+   * @throws Exception
+   */
+  public RunnerApi.Pipeline upgradeTransformsViaTransformService(
+      RunnerApi.Pipeline pipeline, List<String> urnsToOverride, ExternalTranslationOptions options)
+      throws Exception {
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:87]: Avoid throwing generic Exception. (And below.)

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+          TransformT extends org.apache.beam.sdk.transforms.PTransform<InputT, OutputT>>
+      RunnerApi.Pipeline updateTransformViaTransformService(
+          RunnerApi.Pipeline runnerAPIpipeline,
+          String transformId,
+          Endpoints.ApiServiceDescriptor transformServiceEndpoint)
+          throws Exception {
+    PTransform transformToUpgrade =
+        runnerAPIpipeline.getComponents().getTransformsMap().get(transformId);
+    if (transformToUpgrade == null) {
+      throw new Exception("Could not find a transform with the ID " + transformId);
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:160]: Seems like a good candidate for `Preconditions.checkNotNull()`. Or throw something like `IllegalArgumentException`.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 18]
```diff
@@ -0,0 +1,382 @@
...
+    try {
+      configRowSchemaProto =
+          (SchemaApi.Schema)
+              new ObjectInputStream(new ByteArrayInputStream(configRowSchemaBytes.toByteArray()))
+                  .readObject();
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+
+    Row configRow =
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:177]: You already have `configRowBytes`, no need to encode it just to recode it.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 19]
```diff
@@ -0,0 +1,382 @@
...
+            .setComponents(runnerAPIpipeline.getComponents())
+            .setTransform(ptransformBuilder.build())
+            .setNamespace(UPGRADE_NAMESPACE)
+            .build();
+
+    ExpansionApi.ExpansionResponse response =
+        clientFactory.getExpansionServiceClient(transformServiceEndpoint).expand(request);
+
+    if (!Strings.isNullOrEmpty(response.getError())) {
+      throw new IOException(String.format("expansion service error: %s", response.getError()));
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:222]: IOException doesn't seem to be the right thing to throw for a mis-configured transform.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 20]
```diff
@@ -0,0 +1,382 @@
...
+              "Could not find an output with tag "
+                  + entry.getKey()
+                  + " for the transform "
+                  + expandedTransform);
+        }
+        inputReplacements.put(entry.getValue(), newOutput);
+      }
+    }
+
+    String newTransformId = transformId + "_upgraded";
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:274]: We can't guarantee this won't be a collision. On the other hand, there's no need to give it a new name at all, just put the new transform in the map at the old transforms name (and all references will remain valid).

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 21]
```diff
@@ -0,0 +1,382 @@
...
+                              inputReplacements.containsKey(inputEntry.getValue())
+                                  ? inputReplacements.get(inputEntry.getValue())
+                                  : inputEntry.getValue();
+                          updatedInputsMap.put(inputEntry.getKey(), updaterValue);
+                        }
+                        transformBuilder.clearInputs();
+                        transformBuilder.putAllInputs(updatedInputsMap);
+                      }
+
+                      // Fix sub-transforms
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:309]: If you keep the name, you won't have to do this.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 22]
```diff
@@ -0,0 +1,382 @@
...
+                      }
+
+                      return transformBuilder.build();
+                    }));
+
+    newComponentsBuilder.clearTransforms();
+    newComponentsBuilder.putAllTransforms(updatedExpandedTransformMap);
+    newComponentsBuilder.putTransforms(newTransformId, expandedTransform);
+
+    // We fix the root in case the overridden transform was one of the roots.
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:331]: Or this.

**Review by robertwb — COMMENTED** [review]
Thanks, this is looking pretty good. Only some minor comments.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java (hunk 20 above)
Comment (chamikaramj) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java]: Done.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java (hunk 21 above)
Comment (chamikaramj) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java]: Done.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java (hunk 22 above)
Comment (chamikaramj) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java]: Done.

**Review by chamikaramj — COMMENTED** [review]
Thanks. PTAL.

**Comment by chamikaramj** [comment]
Friendly ping :)

**Review by robertwb — APPROVED** [review]
Thanks! This looks good to me.

**Comment by chamikaramj** [comment]
Thanks.
